coderd/wsbuilder/wsbuilder.go

```diff
@@ -0,0 +1,741 @@
+		var pqe *pq.Error
+		if xerrors.As(err, &pqe) {
+			if pqe.Code == "40001" {
+				// serialization error, retry
+				continue
+			}
+		}
+		if err != nil {
+			// Other (hard) error
+			return nil, nil, err
+		}
+		return workspaceBuild, provisionerJob, nil
+	}
+	return nil, nil, xerrors.Errorf("too many errors; last error: %w", err)
+}
+
+// buildTx contains the business logic of computing a new build.  Attributes of the new database objects are computed
+// in a functional style, rather than imperative, to emphasize the logic of how they are defined.  A simple cache
+// of database-fetched objects is stored on the struct to ensure we only fetch things once, even if they are used in
+// the calculation of multiple attributes.
+//
+// In order to utilize this cache, the functions that compute build attributes use a pointer receiver type.
+func (b *Builder) buildTx(authFunc func(action rbac.Action, object rbac.Objecter) bool) (
+	*database.WorkspaceBuild, *database.ProvisionerJob, error,
+) {
+	if authFunc != nil {
+		err := b.authorize(authFunc)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	err := b.checkTemplateVersionMatchesTemplate()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = b.checkTemplateJobStatus()
+	if err != nil {
+		return nil, nil, err
+	}
+	err = b.checkRunningBuild()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	template, err := b.getTemplate()
+	if err != nil {
+		return nil, nil, BuildError{http.StatusInternalServerError, "failed to fetch template", err}
+	}
+
+	templateVersionJob, err := b.getTemplateVersionJob()
+	if err != nil {
+		return nil, nil, BuildError{
+			http.StatusInternalServerError, "failed to fetch template version job", err,
+		}
+	}
+
+	legacyParameters, err := b.getLastParameterValues()
+	if err != nil {
+		return nil, nil, BuildError{
+			http.StatusInternalServerError,
+			"failed to fetch previous legacy parameters.",
+			err,
+		}
+	}
+
+	// if we haven't been told specifically who initiated, default to owner
+	if b.initiator == uuid.Nil {
+		b.initiator = b.workspace.OwnerID
+	}
+	// default reason is initiator
+	if b.reason == "" {
+		b.reason = database.BuildReasonInitiator
+	}
+
+	// Write/Update any new params
+	now := database.Now()
+	for _, param := range b.legacyParameterValues {
+		for _, exists := range legacyParameters {
+			// If the param exists, delete the old param before inserting the new one
+			if exists.Name == param.Name {
+				err = b.store.DeleteParameterValueByID(b.ctx, exists.ID)
+				if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+					return nil, nil, BuildError{
+						http.StatusInternalServerError,
+						fmt.Sprintf("Failed to delete old param %q", exists.Name),
+						err,
+					}
+				}
+			}
+		}
+
+		// If the value is empty, we don't want to save it on database so
+		// Terraform can use the default value
+		if param.SourceValue == "" {
+			continue
+		}
+
+		_, err = b.store.InsertParameterValue(b.ctx, database.InsertParameterValueParams{
```

Comment:
This can also use an errgroup

Reply:
Can we make that a later optimization?  Like if we have some traces or whatever that show this is a bottleneck.

I feel like your typical template has like 5 or fewer parameters, so serializing the calls isn't a big deal.  It's what the code I'm refactoring from does.